DashKitUI/Views/DashSettingsView.swift

```diff
@@ -8,13 +8,90 @@
 
 import SwiftUI
 import LoopKitUI
+import DashKit
 
 struct DashSettingsView<Model>: View where Model: DashSettingsViewModelProtocol  {
     
     @ObservedObject var viewModel: Model
+    
+    @State private var showingDeleteConfirmation = false
+    
+    @Environment(\.horizontalSizeClass) var horizontalSizeClass
+    
+    let dateFormatter: DateFormatter = {
+        let dateFormatter = DateFormatter()
+        dateFormatter.timeStyle = .short
+        dateFormatter.dateStyle = .medium
+        dateFormatter.doesRelativeDateFormatting = true
+        return dateFormatter
+    }()
 
     weak var navigator: DashUINavigator?
     
+    private var daysRemaining: Int? {
+        if case .timeRemaining(let remaining, _, _) = viewModel.lifeState, remaining > .days(1) {
+            return Int(remaining.days)
+        }
+        return nil
+    }
+    
+    private var hoursRemaining: Int? {
+        if case .timeRemaining(let remaining, _, _) = viewModel.lifeState, remaining > .hours(1) {
+            return Int(remaining.hours.truncatingRemainder(dividingBy: 24))
+        }
+        return nil
+    }
+    
+    private var minutesRemaining: Int? {
+        if case .timeRemaining(let remaining, _, _) = viewModel.lifeState, remaining < .hours(2) {
+            return Int(remaining.minutes.truncatingRemainder(dividingBy: 60))
+        }
+        return nil
+    }
+    
+    func timeComponent(value: Int, units: String) -> some View {
+        Group {
+            Text(String(value)).font(.title).fontWeight(.bold)
+            Text(units).foregroundColor(.secondary)
+        }
+    }
+    
+    var lifecycleProgress: some View {
+        VStack(spacing: 10) {
+            HStack(alignment: .lastTextBaseline, spacing: 3) {
+                Text(self.viewModel.lifeState.localizedLabelText)
+                    .foregroundColor(Color(UIColor.secondaryLabel))
+                Spacer()
+                daysRemaining.map { (days) in
+                    timeComponent(value: days, units: days == 1 ? "day" : "days")
+                }
+                hoursRemaining.map { (hours) in
+                    timeComponent(value: hours, units: hours == 1 ? "hour" : "hours")
+                }
+                minutesRemaining.map { (minutes) in
+                    timeComponent(value: minutes, units: minutes == 1 ? "minute" : "minutes")
+                }
+            }
+            ProgressView(progress: CGFloat(self.viewModel.lifeState.progress)).accentColor(self.viewModel.lifeState.progressColor)
+        }
+    }
+    
+    var timeZoneString: String {
```

Comment:
[nit] Hm...seems like a handy utility function that might be worth pulling out into a utility class, and unit tested?  Maybe consider `LoopKit/Extensions/TimeZone.swift`?

Reply:
The DateComponentsFormatter constructor is one of those things that looks innocuous, but actual carries a number of assumptions about current locale, timezone etc, and does a lot of work under the hood, instantiating support for locales, timezones, etc, and can be quite costly. I'm loathe to bury this in another layer that looks trivial.  Typically I recommend that people keep dateformatters at the view/vc/view model layer, and manage their lifecycle in a way that's appropriate for the view.  It's not really much of an issue until someone puts one of these things in a tableview where each row is constructing formatters, but I keep this pattern to help on that front.